File: pong/webassembly/basico/pong.c
#include <emscripten/emscripten.h>
#include <emscripten/html5.h>
#include <stdio.h>

// Variáveis do jogo
float paddleY = 50.0f;
float paddleSpeed = 5.0f;
float ballX = 400.0f;
float ballY = 300.0f;
float ballSpeedX = 5.0f;
float ballSpeedY = 2.0f;

// Função de atualização do jogo
EMSCRIPTEN_KEEPALIVE
void updateGame() {
    // Atualizar a posição da raquete
    paddleY += paddleSpeed;

    // Verificar limites da raquete
    if (paddleY < 0) paddleY = 0;
    if (paddleY > 500) paddleY = 500;

    // Atualizar a posição da bola
    ballX += ballSpeedX;
    ballY += ballSpeedY;

    // Verificar colisão com as paredes verticais
    if (ballY < 0 || ballY > 590) {
        ballSpeedY = -ballSpeedY;
    }

    // Verificar colisão com a raquete
    if (ballX > 780 && ballX < 800 && ballY > paddleY && ballY < paddleY + 100) {
        ballSpeedX = -ballSpeedX;
    }

    // Verificar se a bola ultrapassou a raquete
    if (ballX > 800) {
        // Resetar a posição da bola
        ballX = 400.0f;
        ballY = 300.0f;

        // Mudar a direção da bola
        ballSpeedX = -ballSpeedX;
    }

    // Chamar função JavaScript para renderizar
    char script[50];
    sprintf(script, "renderGame(%f, %f)", ballX, ballY);
    emscripten_run_script(script);
}

// Função principal
int main() {
    // Iniciar o loop de atualização
    emscripten_set_main_loop(updateGame, 60, 1);

    return 0;
}

File: pong/web/chatgpt/simples/pong.js
// pong.js

// Constantes e variáveis do jogo
const gameContainer = document.getElementById("game");
const ball = document.getElementById("ball");
const leftPaddle = document.getElementById("leftPaddle");
const rightPaddle = document.getElementById("rightPaddle");

let ballX = 290;
let ballY = 190;
let ballSpeedX = 3;
let ballSpeedY = 3;
let leftPaddleY = 160;
let rightPaddleY = 160;
const paddleSpeed = 10;

// Função para atualizar a posição da bola
function moveBall() {
  ballX += ballSpeedX;
  ballY += ballSpeedY;

  // Colisão com paredes superior e inferior
  if (ballY >= 385 || ballY <= 0) {
    ballSpeedY = -ballSpeedY;
  }

  // Colisão com as pás
  if (
    ballX <= 25 &&
    ballY >= leftPaddleY &&
    ballY <= leftPaddleY + 80
  ) {
    ballSpeedX = -ballSpeedX;
  }

  if (
    ballX >= 560 &&
    ballY >= rightPaddleY &&
    ballY <= rightPaddleY + 80
  ) {
    ballSpeedX = -ballSpeedX;
  }

  // Se a bola passar das pás
  if (ballX < 0 || ballX > 600) {
    ballX = 290;
    ballY = 190;
    ballSpeedX = -ballSpeedX;
  }

  ball.style.left = ballX + "px";
  ball.style.top = ballY + "px";
}

// Função para mover as pás
function movePaddles() {
  document.addEventListener("keydown", function (event) {
    if (event.key === "ArrowUp" && rightPaddleY > 0) {
      rightPaddleY -= paddleSpeed;
    } else if (event.key === "ArrowDown" && rightPaddleY < 320) {
      rightPaddleY += paddleSpeed;
    }

    if (event.key === "w" && leftPaddleY > 0) {
      leftPaddleY -= paddleSpeed;
    } else if (event.key === "s" && leftPaddleY < 320) {
      leftPaddleY += paddleSpeed;
    }

    leftPaddle.style.top = leftPaddleY + "px";
    rightPaddle.style.top = rightPaddleY + "px";
  });
}

// Função de loop do jogo
function gameLoop() {
  moveBall();
  movePaddles();
  requestAnimationFrame(gameLoop);
}

// Iniciar o loop do jogo
gameLoop();
